packages/ee/billing/backend/usage.service.ts

```diff
@@ -0,0 +1,77 @@
+import { billingService } from "./billing.service";
+import { ActivepiecesError, ErrorCode, FlowVersion, ProjectId, Trigger, Action, apId } from "@activepieces/shared";
+import { databaseConnection } from "@backend/database/database-connection";
+import { createRedisLock } from "@backend/database/redis-connection";
+import { ProjectUsage } from "../shared/usage";
+import { ProjectUsageEntity } from "./usage.entity";
+import { logger } from "@backend/helper/logger";
+import dayjs from "dayjs";
+
+const projectUsageRepo = databaseConnection.getRepository<ProjectUsage>(ProjectUsageEntity);
+
+export const usageService = {
+    async limit(request: { projectId: ProjectId; flowVersion: FlowVersion; }): Promise<{ perform: true }> {
+        const quotaLock = await createRedisLock(5 * 1000);
+        try {
+            quotaLock.acquire(`usage_${request.projectId}}`);
+            const projectUsage = await usageService.getUsage({ projectId: request.projectId });
+            const numberOfSteps = countSteps(request.flowVersion);
+            const projectPlan = await billingService.getPlan({ projectId: request.projectId });
+            if (projectUsage.consumedTasks + numberOfSteps > projectPlan.tasks) {
+                throw new ActivepiecesError({
+                    code: ErrorCode.TASK_QUOTA_EXCEEDED,
+                    params: { id: request.projectId },
+                });
+            }
+            projectUsage.consumedTasks += numberOfSteps;
+            await projectUsageRepo.save(projectUsage);
+        } catch (e) {
+            // TODO add sentry, Don't block for internal error, just continue and log it for now.
+            logger.error(e);
```

Comment:
This catch block renders the `limit` method useless. you probably meant to check if the error is an instance of `ActivepiecesError` or not and act accordingly.

Reply:
Well it's better to not ruin the experience for the users, but I have integrated sentry, so we will be alerted if there is any issues there.